pkg/objstore/client/factory.go

```diff
@@ -35,7 +35,7 @@ var ErrNotFound = errors.New("not found bucket")
 
 // NewBucket initializes and returns new object storage clients.
 // NOTE: confContentYaml can contain secrets.
-func NewBucket(logger log.Logger, confContentYaml []byte, reg *prometheus.Registry, component string) (objstore.Bucket, error) {
+func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registerer, component string) (objstore.Bucket, error) {
```

Comment:
Regarding `*prometheus.Registry` -> `prometheus.Registerer`, Does Registerer have some internal state that it a pointer that will be mutated as `objstore.BucketWithMetrics` is called?  If not, what will `objstore.BucketWithMetrics` do?

Reply:
`Registerer` is an interface that is `Registry` implements. The implementation has its's state indeed but is not mutated pass it in  `objstore.BucketWithMetrics` and that's the point, to all registering new metrics (:  (in fact in as  `prometheus.Registerer` `*prometheus.Registry` is passed)

The reason for change:
* Accept interfaces is right Golang paradigm
* If we accept here a struct pointer `*prometheus.Registry` and `objstore.BucketWithMetrics` accepts interface and we pass `nil` from above (valid case in tests),  `objstore.BucketWithMetrics` attempt on checking `reg != nil` because the given nil is not equals to nil ;p it is equal to `* prometheus.Registry ni)` More details here: https://tour.golang.org/methods/12